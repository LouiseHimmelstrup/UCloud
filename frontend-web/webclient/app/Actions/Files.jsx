import {Cloud} from "../../authentication/SDUCloudObject"
import { RECEIVE_FILES, UPDATE_FILES_PER_PAGE, UPDATE_FILES, SET_LOADING, UPDATE_PATH, TO_PAGE } from "../Reducers/Files"

/*
* Creates a promise to fetch files. Sorts the files based on sorting function passed,
* and implicitely sets @filesLoading to false in the reducer when the files are fetched.
*/
export const fetchFiles = (path, sorting, sortAscending) => 
  Cloud.get(`files?path=${path}`).then((res) => {
      let files = res.response;
      files.forEach(file => file.isChecked = false);
      if (sorting) {
        files = sorting(files, sortAscending);
      }
      return receiveFiles(files);
  });

export const toPage = (pageNumber) => ({
  type: TO_PAGE,
  pageNumber: pageNumber
});

/*
* Updates the files stored. 
* Intended for use when sorting the files, checking or favoriting, for instance.
*/
export const updateFiles = (files) => {
  return {
    type: UPDATE_FILES,
    files
  }
};

export const setLoading = (loading) => {
  return {
    type: SET_LOADING,
    loading
  }
};

export const updatePath = (newPath) => ({
    type: UPDATE_PATH,
    path: newPath
});

export const updateFilesPerPage = (filesPerPage, files) => {
  files.forEach(file => file.isChecked = false);
  return {
    type: UPDATE_FILES_PER_PAGE,
    filesPerPage,
    files
  }
};

const receiveFiles = (files, path) => ({
  type: RECEIVE_FILES,
  files,
  path
});