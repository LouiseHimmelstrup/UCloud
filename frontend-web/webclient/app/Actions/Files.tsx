import { Cloud } from "../../authentication/SDUCloudObject";
import {
    RECEIVE_FILES,
    UPDATE_FILES,
    SET_FILES_LOADING,
    UPDATE_PATH,
    SET_FILES_SORTING_COLUMN,
    FILE_SELECTOR_SHOWN,
    SET_FILE_SELECTOR_LOADING,
    RECEIVE_FILE_SELECTOR_FILES,
    SET_FILE_SELECTOR_CALLBACK,
    SET_DISALLOWED_PATHS,
    SET_CREATING_FOLDER,
    SET_EDITING_FILE,
    RESET_FOLDER_EDITING
} from "../Reducers/Files";
import { sortFilesByTypeAndName, failureNotification } from "../UtilityFunctions";
import { Page, emptyPage, File } from "../types/types";
import { SortOrder, SortBy } from "../SiteComponents/Files/Files";

/**
* Creates a promise to fetch files. Sorts the files based on sorting function passed,
* and implicitly sets {filesLoading} to false in the reducer when the files are fetched.
* @param {string} path is the path of the folder being queried
* @param {number} itemsPerPage number of items to be fetched
* @param {Page<File>} page number of the page to be fetched
*/
export const fetchFiles = (path: string, itemsPerPage: number, page: number, order: SortOrder, sortBy: SortBy) =>
    Cloud.get(`files?path=${path}&itemsPerPage=${itemsPerPage}&page=${page}&order=${order}&sortBy=${sortBy}`).then(({ response }) =>
        receiveFiles(response, path, order, sortBy)
    ).catch(() => {
        failureNotification("An error occurred fetching files for this folder.");
        return receiveFiles(emptyPage, path, order, sortBy);
    });

/**
* Updates the files stored. 
* Intended for use when sorting the files, checking or favoriting, for instance.
* @param {Page<File>} page contains the currently held page with modifications made to the files client-side
*/
export const updateFiles = (page: Page<File>) => ({
    type: UPDATE_FILES,
    page
});

/**
 * Sets whether or not the component is loading
 * @param {boolean} loading 
 */
export const setLoading = (loading: boolean) => ({
    type: SET_FILES_LOADING,
    loading
});

/**
 * Updates the path currently held intended for the files/fileinfo components.
 * @param {string} path - The path to store
 */
export const updatePath = (path: string) => ({
    type: UPDATE_PATH,
    path
});

/**
 * The function used for the actual receiving the files, rather than the promise
 * @param {Page<File>} page - Contains the page
 * @param {string} path - The path the files were retrieved from
 * @param {SortOrder} sortOrder - The order in which the files were sorted
 * @param {SortBy} sortBy - the value the sorting was based on
 */
const receiveFiles = (page: Page<File>, path: string, sortOrder: SortOrder, sortBy: SortBy) => {
    page.items.forEach((f) => f.isChecked = false);
    return {
        type: RECEIVE_FILES,
        page,
        path,
        sortOrder,
        sortBy
    }
};


export const setSortingColumn = (index, name) => ({
    type: SET_FILES_SORTING_COLUMN,
    index,
    name
});

export const fileSelectorShown = (state) => ({
    type: FILE_SELECTOR_SHOWN,
    state
});

export const receiveFileSelectorFiles = (files, path) => ({
    type: RECEIVE_FILE_SELECTOR_FILES,
    files,
    path
});

export const fetchPageFromPath = (path: string, itemsPerPage: number, order: SortOrder, sortBy: SortBy) =>
    Cloud.get(`files/lookup?path=${path}&itemsPerPage=${itemsPerPage}&order=${order}&sortBy=${sortBy}`)
        .then(({ response }) => receiveFiles(response, path, order, sortBy)).catch(() => {
            failureNotification(`An error occured fetching the page for file ${path}`);
            return { type: "ERROR" };
        }
    );

// FIXME add pagination to FileSelector and rewrite this:
export const fetchFileselectorFiles = (path: string, page: number, itemsPerPage: number) =>
    Cloud.get(`files?path=${path}&page=${page}&itemsPerPage=${itemsPerPage}`).then(({ response }) => {
        let files = response.items;
        files.forEach(file => file.isChecked = false);
        files = sortFilesByTypeAndName(files, true);
        return receiveFileSelectorFiles(files, path);
    }).catch(() => {
        failureNotification("An error occurred when fetching files for fileselection");
        return { type: "ERROR" }; // FIXME Will end up in default. Should have case for this
    });

export const setFileSelectorLoading = () => ({
    type: SET_FILE_SELECTOR_LOADING
})

export const setDisallowedPaths = (paths) => ({
    type: SET_DISALLOWED_PATHS,
    paths
});

export const setFileSelectorCallback = (callback) => ({
    type: SET_FILE_SELECTOR_CALLBACK,
    callback
});

export const setEditingFile = (editFileIndex: number) => ({
    type: SET_EDITING_FILE,
    editFileIndex
});

export const setCreatingFolder = (creatingFolder: boolean) => ({
    type: SET_CREATING_FOLDER,
    creatingFolder
});

export const resetFolderEditing = () => ({
    type: RESET_FOLDER_EDITING
})