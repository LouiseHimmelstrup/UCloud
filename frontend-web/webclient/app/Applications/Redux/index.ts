import * as ViewObject from "./ViewObject";
import * as ViewReducer from "./ViewReducer";
import * as ApplicationObject from "./ApplicationsObject";
import * as ApplicationReducer from "./ApplicationsReducer";

export type Reducers = ViewReducer.Reducer & ApplicationReducer.Reducer;
export type Objects = ViewObject.Wrapper & ApplicationObject.Wrapper;

export function init(): Objects {
    return { 
        ...ViewObject.init(),
        ...ApplicationObject.init()
    }
}

export const reducers: Reducers = {
    applicationView: ViewReducer.default,
    applicationsBrowse: ApplicationReducer.default
};