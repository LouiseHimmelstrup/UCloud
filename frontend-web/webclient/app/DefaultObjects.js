import {Cloud} from "../authentication/SDUCloudObject"
import {tusConfig} from "./Configurations";
import Uppy from "uppy";

export const DefaultStatus = {
    title: "No Issues",
    level: "NO ISSUES",
    body: "The system is running as intended.",
};

export const RightsMap = {
    NONE: 0,
    READ: 1,
    READ_WRITE: 2,
    OWN: 3,
};

export const AnalysesStatusMap = {
    "PENDING": 0,
    "IN PROGRESS": 1,
    "COMPLETED": 2,
};

export const RightsNameMap = {
    NONE: "None",
    READ: "Read",
    READ_WRITE: "Read/Write",
    OWN: "Own",
};

export const SensitivityLevel = {
    "OPEN_ACCESS": "Open Access",
    "CONFIDENTIAL": "Confidential",
    "SENSITIVE": "Sensitive",
};

export const SensitivityLevelMap = {
    "OPEN_ACCESS": 0,
    "CONFIDENTIAL": 1,
    "SENSITIVE": 2,
};

const initializeUppy = (restrictions) => 
    Uppy.Core({
        autoProceed: false,
        debug: false,
        restrictions: restrictions,
        meta: {
            sensitive: false,
        },
        onBeforeUpload: () => {
            return Cloud.receiveAccessTokenOrRefreshIt().then((data) => {
                tusConfig.headers["Authorization"] = `Bearer ${data}`;
            });
        }
    }).use(Uppy.Tus, tusConfig);

export const initObject = {
    files: {
        files: [],
        filesShown: 10,
        filesPerPage: 10,
        currentFilesPage: 0,
        loading: false,
        path: "",
        filesInfoPath: "",
        filesInfoFile: null,
        projects: []
    },
    uppy: {
        uppyFiles: initializeUppy({maxNumberOfFiles: false}),
        uppyFilesOpen: false,
        uppyRunApp: initializeUppy({ maxNumberOfFiles: 1}),
        uppyRunAppOpen: false
    },
    status: {
        status: DefaultStatus,
        title: ""
    }
};