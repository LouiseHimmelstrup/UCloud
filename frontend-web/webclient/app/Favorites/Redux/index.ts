import * as FavoritesObject from "./FavoriteObject";
import * as FavoritesReducer from "./FavoritesReducer";

export type Reducers = FavoritesReducer.Reducer;
export type Objects = FavoritesObject.Wrapper;

export function init(): Objects {
    return {
        ...FavoritesObject.init()
    }
}

export const reducers: Reducers = {
    favorites: FavoritesReducer.default
}