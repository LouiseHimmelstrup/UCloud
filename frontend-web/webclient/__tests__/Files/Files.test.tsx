import * as React from "react";
import * as Renderer from "react-test-renderer";
import { emptyPage, KeyCode } from "DefaultObjects";
import { updateFiles } from "Files/Redux/FilesActions";
import Files, { FilesTable, FileOperations } from "Files/Files";
import { DropdownItem } from "semantic-ui-react";
import { setLoading } from "Files/Redux/FilesActions"
import { SortOrder, SortBy, Operation, PredicatedOperation, File } from "Files";
import { mockFiles_SensitivityConfidential } from "../mock/Files"
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { createMemoryHistory } from "history";
import files from "Files/Redux/FilesReducer";
import { Button, Dropdown } from "semantic-ui-react";
import { AllFileOperations } from "Utilities/FileUtilities";
import { configureStore } from "Utilities/ReduxUtilities";
import { initFiles } from "DefaultObjects";
import { configure, mount } from "enzyme";
import { Page } from "Types";
import * as Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });


const createMockStore = (filesPage?: Page<File>) => {
    const store = configureStore({ files: initFiles({ homeFolder: "/home/user@test.abc/" }) }, { files });
    if (!!filesPage) {
        store.dispatch(updateFiles(filesPage));
    }
    return store;
}

const mockHistory = createMemoryHistory();
const nullOp = () => null;

const fileOperations = AllFileOperations(true, {
    setFileSelectorCallback: nullOp,
    showFileSelector: nullOp,
    setDisallowedPaths: nullOp,
    fetchPageFromPath: nullOp
}, nullOp, mockHistory);

describe("FilesTable", () => {
    test("Render empty", () => {
        expect(Renderer.create(
            <FilesTable
                files={emptyPage.items}
                fileOperations={[]}
                sortOrder={SortOrder.ASCENDING}
                sortingColumns={[SortBy.PATH, SortBy.MODIFIED_AT]}
                sortFiles={() => null}
                onCheckFile={() => null}
                refetchFiles={() => null}
                sortBy={SortBy.PATH}
                onFavoriteFile={() => null}
            />
        ).toJSON()).toMatchSnapshot();
    });

    test("Render non-empty", () => {
        expect(Renderer.create(
            <MemoryRouter>
                <FilesTable
                    files={mockFiles_SensitivityConfidential.items}
                    fileOperations={[]}
                    sortOrder={SortOrder.ASCENDING}
                    sortingColumns={[SortBy.PATH, SortBy.MODIFIED_AT]}
                    sortFiles={() => null}
                    onCheckFile={() => null}
                    refetchFiles={() => null}
                    sortBy={SortBy.PATH}
                    onFavoriteFile={() => null}
                />
            </MemoryRouter>
        ).toJSON()).toMatchSnapshot()
    })
});

describe("Files-component", () => {
    test("Full Files component, no files", () => {
        expect(Renderer.create(
            <Provider store={createMockStore()}>
                <MemoryRouter>
                    <Files
                        history={mockHistory}
                        match={{ params: [], isExact: false, path: "", url: "home" }}
                    />
                </MemoryRouter>
            </Provider>).toJSON()
        ).toMatchSnapshot();
    });

    test("Full Files component, full page of files", () => {
        expect(Renderer.create(
            <Provider store={createMockStore(mockFiles_SensitivityConfidential)}>
                <MemoryRouter>
                    <Files
                        history={mockHistory}
                        match={{ params: [], isExact: false, path: "", url: "home" }}
                    />
                </MemoryRouter>
            </Provider>).toJSON()
        ).toMatchSnapshot();
    });
});


describe("File operations", () => {
    test("Empty files list, button, empty FilesOperations", () => {
        expect(Renderer.create(
            <FileOperations
                fileOperations={[]}
                files={createMockStore().getState().files.page.items}
                As={Button}
                fluid
                basic
            />
        ).toJSON()).toMatchSnapshot()
    });

    test("Empty files list, dropdown.item, empty FilesOperations", () =>
        expect(Renderer.create(
            <FileOperations
                fileOperations={[]}
                files={createMockStore().getState().files.page.items}
                As={Dropdown.Item}
            />
        ).toJSON()).toMatchSnapshot()
    );

    test("Files list with items, button, empty FilesOperations", () =>
        expect(Renderer.create(
            <FileOperations
                fileOperations={[]}
                files={createMockStore(mockFiles_SensitivityConfidential).getState().files.page.items}
                As={Button}
                fluid
                basic
            />
        ).toJSON()).toMatchSnapshot()
    );

    test("Files list with items, dropdown.item, empty FilesOperations", () =>
        expect(Renderer.create(
            <FileOperations
                fileOperations={[]}
                files={createMockStore(mockFiles_SensitivityConfidential).getState().files.page.items}
                As={Button}
                fluid
                basic
            />
        ).toJSON()).toMatchSnapshot()
    );

    test("Empty files list, button, some fileoperations", () =>
        expect(Renderer.create(
            <FileOperations
                fileOperations={fileOperations}
                files={createMockStore().getState().files.page.items}
                As={Button}
                fluid
                basic
            />
        )).toMatchSnapshot()
    );

    test("Empty files list, dropdown.item, some fileoperations", () =>
        expect(Renderer.create(
            <FileOperations
                fileOperations={fileOperations}
                files={createMockStore().getState().files.page.items}
                As={Dropdown.Item}
                fluid
                basic
            />
        )).toMatchSnapshot()
    );

    test("Files list with items, button, some fileoperations", () =>
        expect(Renderer.create(
            <FileOperations
                fileOperations={fileOperations}
                files={createMockStore(mockFiles_SensitivityConfidential).getState().files.page.items}
                As={Button}
                fluid
                basic
            />
        )).toMatchSnapshot()
    );

    test("Files list with items, dropdown.item, some fileoperations", () =>
        expect(Renderer.create(
            <FileOperations
                fileOperations={fileOperations}
                files={createMockStore(mockFiles_SensitivityConfidential).getState().files.page.items}
                As={Dropdown.Item}
            />
        )).toMatchSnapshot()
    );
});


const getOperationFrom = (node: Renderer.ReactTestRenderer, operation: string) =>
    node.root.findAllByType(FileOperations)[0].props.fileOperations.find((it: Operation) => it.text === operation);

const getPredicatedOperationFrom = (node: Renderer.ReactTestRenderer, operation: string, onTrue: boolean) =>
    node.root.findAllByType(FileOperations)[0].props.fileOperations.filter(it => it.onTrue).find((it: PredicatedOperation) => it[onTrue ? "onTrue" : "onFalse"].text === operation);

describe("FilesTable Operations being mounted", () => {
    // Non-predicated operations
    ["Share", "Download", "Copy", "Move", "Delete", "Properties"].forEach(operation =>
        test(`${operation} is rendered as dropdown`, function () {
            const node = Renderer.create(
                <Provider store={createMockStore(mockFiles_SensitivityConfidential)}>
                    <MemoryRouter>
                        <Files
                            history={mockHistory}
                            match={{ params: [], isExact: false, path: "", url: "home" }}
                        />
                    </MemoryRouter>
                </Provider>)
            const op: Operation = getOperationFrom(node, operation);
            const toMatch = (fileOperations.filter((it: Operation) => it.text === operation)[0] as Operation);
            // FIXME Currently needed due to dynamic nature of tests;
            expect(JSON.parse(JSON.stringify(op))).toEqual(JSON.parse(JSON.stringify(toMatch)));
        }));

    test("Rename", () => {
        const operationName = "Rename";
        const table = Renderer.create(
            <Provider store={createMockStore(mockFiles_SensitivityConfidential)}>
                <MemoryRouter>
                    <Files
                        history={mockHistory}
                        match={{ params: [], isExact: false, path: "", url: "home" }}
                    />
                </MemoryRouter>
            </Provider>);
        const operation = getOperationFrom(table, operationName);
        expect(operation).toBeDefined()
    });

    test("Create Project", () => {
        const operationName = "Create Project";
        const table = Renderer.create(
            <MemoryRouter>
                <FilesTable
                    files={createMockStore(mockFiles_SensitivityConfidential).getState().files.page.items}
                    fileOperations={fileOperations}
                    sortOrder={SortOrder.ASCENDING}
                    sortingColumns={[SortBy.PATH, SortBy.MODIFIED_AT]}
                    sortFiles={() => null}
                    onCheckFile={() => null}
                    refetchFiles={() => null}
                    sortBy={SortBy.PATH}
                    onFavoriteFile={() => null}
                />
            </MemoryRouter>);
        const operation = getPredicatedOperationFrom(table, operationName, false);
        expect(operation).toBe(fileOperations.filter((it: PredicatedOperation) => it.onTrue).find((it: PredicatedOperation) => it["onFalse"].text === operationName));
    });

    test("Edit Project", () => {
        const operationName = "Edit Project";
        const table = Renderer.create(
            <MemoryRouter>
                <FilesTable
                    files={createMockStore(mockFiles_SensitivityConfidential).getState().files.page.items}
                    fileOperations={fileOperations}
                    sortOrder={SortOrder.ASCENDING}
                    sortingColumns={[SortBy.PATH, SortBy.MODIFIED_AT]}
                    sortFiles={() => null}
                    onCheckFile={() => null}
                    refetchFiles={() => null}
                    sortBy={SortBy.PATH}
                    onFavoriteFile={() => null}
                />
            </MemoryRouter>);
        const operation = getPredicatedOperationFrom(table, operationName, true);
        expect(operation).toBe(fileOperations.filter((it: PredicatedOperation) => it.onTrue).find((it: Operation) => it["onTrue"].text === operationName));
    });
});

describe("Files components usage", () => {

    test("Start creation of folder", () => {
        const emptyPageStore = createMockStore();
        let files = mount(
            <Provider store={emptyPageStore}>
                <MemoryRouter>
                    <Files
                        history={mockHistory}
                        match={{ params: ["/else/thing"], isExact: false, path: "/This/Is/A/Path", url: "home" }}
                    />
                </MemoryRouter>
            </Provider>);
        expect(emptyPageStore.getState().files.page.items.length).toBe(0);
        files.findWhere(it => it.props().content === "New folder").simulate("click");
        // Test cleaning up other mock folders
        files.findWhere(it => it.props().content === "New folder").simulate("click");
        emptyPageStore.dispatch(setLoading(false));
        files = files.update();
        expect(emptyPageStore.getState().files.page.items.length).toBe(1);
        files.findWhere(it => it.props().className === "ui red small basic button").simulate("click");
        expect(emptyPageStore.getState().files.page.items.every(it => !it.isMockFolder));
    });
});

describe("FilesTable Operations being used", () => {
    test("Start and stop renaming", () => {
        const fullPageStore = createMockStore(mockFiles_SensitivityConfidential);
        let node = mount(
            <Provider store={fullPageStore}>
                <MemoryRouter>
                    <Files
                        history={mockHistory}
                        match={{ params: [], isExact: false, path: "/This/Is/A/Path", url: "home" }}
                    />
                </MemoryRouter>
            </Provider>);
        const firstBeingRenamedCount = fullPageStore.getState().files.page.items.filter(it => it.beingRenamed).length;
        expect(firstBeingRenamedCount).toBe(0);
        node.find(DropdownItem).findWhere(it => it.type() === "span").findWhere(it => it.text() === "Rename").first().simulate("click");
        // FIXME Must set loading as false as the component tries to fetch new page, I think
        fullPageStore.dispatch(setLoading(false));

        expect(fullPageStore.getState().files.page.items.filter(it => it.beingRenamed).length).toBe(1);
        node = node.update();

        node.find("input").findWhere(it => it.props().type === "text").simulate("keydown", {
            keyCode: KeyCode.ESC,
            target: { value: "New folder Name" }
        });

        node = node.update();

        expect(fullPageStore.getState().files.page.items.filter(it => it.beingRenamed).length).toBe(0);
    });
});