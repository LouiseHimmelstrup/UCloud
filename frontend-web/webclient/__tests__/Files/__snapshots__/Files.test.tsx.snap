// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mount Files component 1`] = `
<div
  className="sc-htpNat gDfBLd"
>
  <div
    className="sc-htpNat sc-bxivhb sc-kjoXOD cQydld"
    height={96}
    width={1}
  >
    <div
      className="sc-bdVaJa fuBZyM"
    >
      <div
        className="sc-htpNat exYHbV"
      >
        <svg
          className="sc-htoDjs fTQtPU"
          cursor="pointer"
          fill="currentcolor"
          height="30px"
          onClick={[Function]}
          viewBox="0 0 148.167 127.257"
          width="30px"
        >
          <path
            d="M0 63.5l10.583 10.583 63.5-52.916 63.5 52.916L148.167 63.5 74.083 0z"
          />
          <path
            d="M74.083 28.149l-50.27 41.618v57.49h34.316V92.312c0-4.162 2.06-11.64 15.954-11.628 13.894.012 15.958 7.406 15.958 11.71v34.863h34.313v-57.49zM124.354 38.92l-19.553-17.224V5.252l19.553.04z"
          />
        </svg>
        <div
          className="sc-jTzLTM ivwLgz"
          cursor="pointer"
          fontSize="11px"
          onClick={[Function]}
        >
          Go to home
        </div>
      </div>
      <div
        className="sc-jTzLTM jFNMAO"
        cursor="inherit"
        fontSize="25px"
      >
        |
      </div>
      <ul
        className="sc-cIShpX dtxAhc"
      >
        <li
          title="Home of jonas@hinchely.dk"
        >
          Home of jonas@hinche
          ...
        </li>
      </ul>
      <div
        className="sc-htpNat btiFIV"
      />
      <div
        className="sc-hzDkRC isPhWx"
        data-tag="dropdown"
      >
        <span
          className="sc-jTzLTM bjceot"
          cursor="pointer"
          onClick={[Function]}
        >
          <span
            className="sc-jTzLTM bnnwzT"
            cursor="inherit"
          >
             
            Files per page 25
          </span>
          <svg
            className="sc-htoDjs krWwdJ"
            clipRule="evenodd"
            cursor="inherit"
            fill="currentcolor"
            fillRule="evenodd"
            height=".7em"
            ml=".7em"
            viewBox="0 0 24 12"
            width=".7em"
          >
            <path
              d="M0 0l11.959 12 12-12H0z"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
  <div
    className="sc-htpNat sc-jKJlTe lopkwb"
  >
    <div
      className="sc-htpNat sc-bxivhb sc-cHGsZl iIOJRB"
      height="100%"
      width={240}
    >
      <div
        className="sc-htpNat eRCWhu"
        height="calc(100% - 20px)"
      >
        <div
          className="sc-bdVaJa sc-bRBYWo gaZbtS"
        >
          <button
            className="sc-bZQynM cZIxad"
            color="blue"
            cursor="pointer"
            onClick={[Function]}
          >
            <span>
              Upload Files
            </span>
          </button>
          <button
            className="sc-bZQynM sc-gzVnrw hHlzxg"
            color="blue"
            cursor="pointer"
            onClick={[Function]}
          >
            <span>
              New Folder
            </span>
          </button>
          <div
            className="sc-htpNat gsvqod"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-htpNat gfqUsZ"
  >
    <div
      className="sc-htpNat sc-jKJlTe fyzMpK"
    >
      <div
        className="sc-htpNat eRCWhu"
        height="calc(100% - 20px)"
      >
        <div
          className="sc-bdVaJa sc-bRBYWo gaZbtS"
        >
          <button
            className="sc-bZQynM cZIxad"
            color="blue"
            cursor="pointer"
            onClick={[Function]}
          >
            <span>
              Upload Files
            </span>
          </button>
          <button
            className="sc-bZQynM sc-gzVnrw hHlzxg"
            color="blue"
            cursor="pointer"
            onClick={[Function]}
          >
            <span>
              New Folder
            </span>
          </button>
          <div
            className="sc-htpNat gsvqod"
          />
        </div>
      </div>
    </div>
    <h3
      className="sc-jTzLTM iUcSta"
      cursor="inherit"
      fontSize={4}
    >
      No files in current folder
    </h3>
  </div>
</div>
`;
